graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/NodeMetadataSocialProofGenerator.java

```diff
@@ -0,0 +1,183 @@
+  private final Int2ObjectMap<Byte2ObjectMap<Long2ObjectMap<LongSet>>> socialProofs;
+  // NodeMetadata (Int) -> Sum of social proof edges (Double)
+  private final Int2DoubleMap socialProofWeights;
+  protected RecommendationType recommendationType;
+  protected IDMask idMask;
+
+  public NodeMetadataSocialProofGenerator(
+    NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph graph,
+    TweetIDMask mask,
+    RecommendationType recommendationType
+  ) {
+    this.graph = graph;
+    this.idMask = mask;
+    this.recommendationType = recommendationType;
+    // Variables such as socialProofs and socialProofWeights, are re-used for each request.
+    // We choose to use an OpenHashMap to future proof the potential performance degradation,
+    // and the memory is re-used, so we are less concerned about the memory overhead of OpenHashMap
+    // compared with ArrayMap.
+    this.socialProofs = new Int2ObjectOpenHashMap<>();
+    this.socialProofWeights = new Int2DoubleOpenHashMap();
+  }
+
+  private void updateSocialProofWeight(int metadataId, double weight) {
+    // We sum the weights of incoming leftNodes as the weight of the rightNode.
+    socialProofWeights.put(
+      metadataId,
+      weight + socialProofWeights.get(metadataId)
+    );
+  }
+
+  private void addSocialProof(int metadataId, byte edgeType, long leftNode, long rightNode) {
+    if (!socialProofs.containsKey(metadataId)) {
+      // We choose ArrayMap over OpenHashMap since the request will have at most a few edge types,
+      // usually less than five.
+      socialProofs.put(metadataId, new Byte2ObjectArrayMap<>());
+      socialProofWeights.put(metadataId, 0);
+    }
+    Byte2ObjectMap<Long2ObjectMap<LongSet>> socialProofMap = socialProofs.get(metadataId);
+
+    // Get the user to tweets map variable by the engagement type.
+    if (!socialProofMap.containsKey(edgeType)) {
+      // We choose to use an OpenHashMap since a single edge type may have dozens or even hundreds
+      // of user seed ids associated.
+      socialProofMap.put(edgeType, new Long2ObjectOpenHashMap<>());
+    }
+    Long2ObjectMap<LongSet> userToTweetsMap = socialProofMap.get(edgeType);
+
+    // Add the connecting user to the user map.
+    if (!userToTweetsMap.containsKey(leftNode)) {
+      // We choose to use an OpenHashSet since a single user may engage with dozens or even
+      // hundreds of tweet ids.
+      userToTweetsMap.put(leftNode, new LongOpenHashSet());
+    }
+    LongSet connectingTweets = userToTweetsMap.get(leftNode);
+
+    // Add the connecting tweet to the tweet set.
+    if (!connectingTweets.contains(rightNode)) {
+      connectingTweets.add(rightNode);
+    }
+  }
+
+  /**
+   * Collect social proofs for a given {@link SocialProofRequest}.
+   *
+   * @param request contains a set of input metadata ids and a set of seed users.
+   */
+  private void collectRecommendations(NodeMetadataSocialProofRequest request) {
+    socialProofs.clear();
```

Comment:
it is cleaner to have a reset function including those two clear functions, and it is better to call the reset function in computeRecommendations right before collectRecommendations. 

could you fix the other TweetSocialProof Generator as well?